compiler/qsc_rca/src/common.rs

```diff
@@ -0,0 +1,354 @@
+    callable: &CallableDecl,
+    pats: &IndexMap<PatId, Pat>,
+) -> Vec<InputParam> {
+    let input_elements = derive_callable_input_pattern_elements(callable, pats);
+    let mut input_params = Vec::new();
+    let mut param_index = InputParamIndex(0);
+    for element in input_elements {
+        let maybe_input_param = match &element.kind {
+            InputPatternElementKind::Discard => Some(InputParam {
+                index: param_index,
+                pat: element.pat,
+                ty: element.ty.clone(),
+                node: None,
+            }),
+            InputPatternElementKind::Node(node_id) => Some(InputParam {
+                index: param_index,
+                pat: element.pat,
+                ty: element.ty.clone(),
+                node: Some(*node_id),
+            }),
+            InputPatternElementKind::Tuple(_) => None,
+        };
+
+        if let Some(input_param) = maybe_input_param {
+            input_params.push(input_param);
+            param_index.0 += 1;
+        }
+    }
+
+    input_params
+}
+
+/// A represenation of a local symbol.
+#[derive(Clone, Debug)]
+pub struct Local {
+    pub node: NodeId,
+    pub pat: PatId,
+    pub ty: Ty,
+    pub kind: LocalKind,
+}
+
+/// Kinds of local symbols.
+#[derive(Clone, Debug)]
+pub enum LocalKind {
+    /// An input parameter with its associated index.
+    InputParam(InputParamIndex),
+    /// A specialization input (i.e. control qubits).
+    SpecInput,
+    /// An immutable binding with the expression associated to it.
+    Immutable(ExprId),
+    /// A mutable binding.
+    Mutable,
+}
+
+pub trait LocalsLookup {
+    fn find(&self, node_id: NodeId) -> Option<&Local>;
+
+    fn get(&self, node_id: NodeId) -> &Local {
+        self.find(node_id).expect("local should exist")
+    }
+}
+
+impl LocalsLookup for FxHashMap<NodeId, Local> {
+    fn find(&self, node_id: NodeId) -> Option<&Local> {
+        self.get(&node_id)
+    }
+}
+
+pub fn initalize_locals_map(input_params: &Vec<InputParam>) -> FxHashMap<NodeId, Local> {
+    let mut locals_map = FxHashMap::<NodeId, Local>::default();
+    for param in input_params {
+        if let Some(node) = param.node {
+            locals_map.insert(
+                node,
+                Local {
+                    node,
+                    pat: param.pat,
+                    ty: param.ty.clone(),
+                    kind: LocalKind::InputParam(param.index),
+                },
+            );
+        }
+    }
+    locals_map
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct GlobalSpecId {
+    pub callable: StoreItemId,
+    pub spec_kind: SpecKind,
+}
+
+impl From<(StoreItemId, SpecKind)> for GlobalSpecId {
+    fn from(value: (StoreItemId, SpecKind)) -> Self {
+        Self {
+            callable: value.0,
+            spec_kind: value.1,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum SpecKind {
+    Body,
+    Adj,
+    Ctl,
+    CtlAdj,
+}
```

Comment:
Should FunctorSetValue be used here. We are operating on HIR and these values are defined already for HIR.

Reply:
Yes. We should move it into a shared location and use it as you suggested for `FunctorApp`.

Reply:
Well, `FunctorSetValue` probably does not need to be moved.

Reply:
Updated to use `FunctorSetValue`.